test/code/providers/config_scanner_test.cpp

```diff
@@ -0,0 +1,669 @@
+    CPPUNIT_TEST( testCnfScan_IncludeDirWithNonCnfFiles_IgnoreNonCnf );
+    CPPUNIT_TEST( testCnfScan_DirectivesFormCycles);
+    CPPUNIT_TEST( testCnfScan_CnfFileMissing_ReturnsFalse);
+    CPPUNIT_TEST( testCnfScan_InvalidDirectives_Ignored );
+
+    CPPUNIT_TEST_SUITE_END();
+
+private:
+
+public:
+    void setUp(void)
+    {
+    }
+
+    void tearDown(void)
+    {
+    }
+
+    void ValidateAuth(std::map< unsigned int, std::string >& expected, MySQL_Authentication& auth)
+    {
+        std::vector<unsigned int> authPorts;
+        auth.GetPortList( authPorts );
+        CPPUNIT_ASSERT_EQUAL(expected.size(), authPorts.size());
+        for (std::vector<unsigned int>::const_iterator it = authPorts.begin(); it != authPorts.end(); ++it)
+        {
+            MySQL_AuthenticationEntry authEntry;
+            auth.GetEntry(*it, authEntry);
+            CPPUNIT_ASSERT_EQUAL(expected[authEntry.port], authEntry.binding);
+        }
+    }
+
+    void CreateConfig(const std::wstring& binding, unsigned int port, bool withMySqld, std::vector< std::wstring >& conf)
+    {
+        std::wstringstream out;
+        conf.push_back(out.str());
+        conf.push_back(L"# A Comment line");
+        conf.push_back(L"[client]");
+        out.str(L"");
+        out << L"port     = " << port ;
+        conf.push_back(out.str());
+        conf.push_back(L"socket		= /var/run/mysqld/mysqld.sock");
+        conf.push_back(L"An irrelevant line without equalsTo");
+        conf.push_back(L"A line = with more than one =  ");
+        out.str(L"");
+        out << L"bind-address   = " << binding;
+        conf.push_back(out.str());
+        conf.push_back(L"");
+        if (withMySqld)
+        {
+            conf.push_back(L"[mysqld]");
+            conf.push_back(L"#");
+            conf.push_back(L"# * Basic Settings");
+            conf.push_back(L"#");
+            conf.push_back(L"user		= mysql");
+            conf.push_back(L"pid-file	= /var/run/mysqld/mysqld.pid");
+            conf.push_back(L"socket		= /var/run/mysqld/mysqld.sock");
+            conf.push_back(L"port = 3306");
+            conf.push_back(L"basedir		= /usr");
+            conf.push_back(L"datadir		= /var/lib/mysql");
+            out.str(L"");
+        }
+        conf.push_back(out.str());
+    }
+
+    void testCnfScan_NoDirectives()
+    {
+        MySQL_Authentication auth( SCXHandle<MySQL_AuthenticationDependencies>(new MySQL_ScannerTestAuthenticationDependencies()));
+        auth.Load();
+        std::map< std::wstring, std::vector< std::wstring > > fileContentsMap;
+        std::map< std::wstring, std::vector< std::wstring > > directoryMap;
+        std::map< unsigned int, std::string > portBindingMap;
+        std::vector < std::wstring > contents;
+
+        std::wstring cnfFileName = L"NoDirectives.cnf";
+        std::wstring bindAddress = L"172.22.13.1";
+        unsigned int port = 3505;
+        CreateConfig(bindAddress, port, true, contents);
```

Comment:
I'm trying hard to avoid a "not invented here" syndrome, and I don't think I am. But I'm not thrilled with the way that you handled this. I sort of expected something similar to the tests that GetLinuxOS has, in that we check in the real-live configuration files on the machine (and I suppose we could check in our test onesas well). It's easy to extend that, you have a bunch of "real-life" examples of actual config files that ship with various systems (like 10 or 12 of them).

Here's what I'm sort of thinking I'd like to see in the unit tests:

1. You create a directory (like test/code/...), that contains support files for various platforms. Again, see what was done for GetLinuxOS. You can check in files from numerous systems, as well as creating directories and files for specific test cases that you're interested in,

2. You get rid of CreateConfig and just read in those files (giving the path in a dependency object, and that path gets prepended to any file that you try and deal with, so something like /etc/mycnf/foo.cnf becomes /users/jeffcof/dev/bld-mysql/mysql/test/code/.../etc/mycnf/foo.cnf). This way, you have just one or two lines of changes; otherwise, we're reading in real-live config files from real-live platforms (as well as your specific tests that you may include),

3. Have your test dependency object allow "flags". So, for example, the production code may detect a loop and call "LoopDectected" in the dependency object. The production version of that may throw an exception, where the test version of that may just set a flag in the dependency object and return. Then, in the test itself, you can be sure the flag was set (i.e. a loop was detected).

In particular, I think #2 would make the tests a lot shorter and easier to extend. You just point a directory for a test, load it, and then look at the resulting object.

I very much like that you're validating specific authentications for specific files. Keep that, that's just what I would have done, and insures that the state you expect is just as anticipated after a set of configuration files are read.

Let me know what you think of this. Again, I do NOT want to have a "not invented here" syndrome, so if you really feel like this approach is better, please tell me why and let's chat.

In the case of GetLinuxOS, we wrote it initially, and then each time we changed the script due to a new platform, I added the actual config files for that platform. Makes sure that the new platform works, all older platforms work, and that it's easily extendable for new stuff down the road.